include/utility/dynarray.h

```diff
@@ -0,0 +1,364 @@
+/************************
+ * @file dynarray.h
+ * @author Zhile Zhu (zhuzhile08@gmail.com)
+ * 
+ * @brief a (slightly stupid) array masked as a vector
+ * 
+ * @date 2023-01-14
+ * 
+ * @copyright Copyright (c) 2023
+ ************************/
+
+#pragma once
+
+#include <type_traits>
+#include <span>
+
+#include <utility/array.h>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
+
```

Comment:
Boom; include both array.h and dynarray.h and we get define clashes. Also, C++ headers should be `.hpp`.

Reply:
Alright, as mentioned above, I'll make that central header and change it all to `.hpp`, then I think this should be good to go.